File: grooves-api/src/main/groovy/com/github/rahulsom/grooves/api/snapshots/TemporalJoin.java
package com.github.rahulsom.grooves.api.snapshots;

import com.github.rahulsom.grooves.api.AggregateType;
import com.github.rahulsom.grooves.api.events.BaseEvent;
import com.github.rahulsom.grooves.api.snapshots.internal.BaseJoin;

/**
 * @param <Aggregate>             The Aggregate this join represents
 * @param <JoinIdType>            The type for the join's {@link #getId()} field
 * @param <JoinedAggregateIdType> The type for the other aggregate that {@link Aggregate} joins to
 * @param <EventIdType>           The type for the {@link EventType}'s id field
 * @param <EventType>             The base type for events that apply to {@link Aggregate}
 * @author Rahul Somasunderam
 */
public interface TemporalJoin<
        Aggregate extends AggregateType,
        JoinIdType, JoinedAggregateIdType,
        EventIdType,
        EventType extends BaseEvent<Aggregate, EventIdType, EventType>>
        extends
        TemporalSnapshot<Aggregate, JoinIdType, EventIdType, EventType>,
        BaseJoin<Aggregate, JoinIdType, JoinedAggregateIdType, EventIdType, EventType> {

}

File: grooves-types/src/main/kotlin/com/github/rahulsom/grooves/api/snapshots/TemporalJoin.kt
package com.github.rahulsom.grooves.api.snapshots

import com.github.rahulsom.grooves.api.AggregateType
import com.github.rahulsom.grooves.api.events.BaseEvent
import com.github.rahulsom.grooves.api.snapshots.internal.BaseJoin

/**
 * A special kind of [TemporalSnapshot] that stores information about joined entities.
 *
 * @param [AggregateIdT]       The type of [AggregateT.id]
 * @param [AggregateT]         The Aggregate this snapshot works over
 * @param [JoinIdT]            The type for [BaseJoin.id]
 * @param [JoinedAggregateIdT] The type for the id of the aggregate that [AggregateT] to
 * @param [EventIdT]           The type for [EventT.id]
 * @param [EventT]             The base type for events that apply to [AggregateT]
 *
 * @author Rahul Somasunderam
 */
interface TemporalJoin<
        AggregateIdT,
        AggregateT : AggregateType<AggregateIdT>,
        JoinIdT,
        JoinedAggregateIdT,
        EventIdT,
        in EventT : BaseEvent<AggregateIdT, AggregateT, EventIdT, in EventT>> :
        TemporalSnapshot<AggregateIdT, AggregateT, JoinIdT, EventIdT, EventT>,
        BaseJoin<AggregateIdT, AggregateT, JoinIdT, JoinedAggregateIdT, EventIdT, EventT>
